File: ios/source/middleware/ResponseMiddleware.swift
import Foundation

class ResponseMiddleware: Middleware {
    var dispatch: Dispatch?

    func execute(state: AppState, action: Action) {
        if case .receivedResponse(let response) = action {
            onResponse(response)
        }
    }

    private func onResponse(_ response: Bloop_Response) {
        if response.hasProject {
            self.dispatch?(.setProject(response.project))
        }

        if response.hasPlaybackState {
            self.dispatch?(.setPlaybackState(response.playbackState))
        }

        if response.hasProgress {
            self.dispatch?(.setProgress(response.progress))
        }

        if !response.projects.isEmpty {
            self.dispatch?(.setProjects(response.projects))
        }

        if !response.cloudProjects.isEmpty {
            self.dispatch?(.setCloudProjects(response.cloudProjects))
        }

        if !response.error.isEmpty {
            self.dispatch?(.addError(response.error))
        }

        if response.hasWaveform {
            let waveform = response.waveform
            let action = Action.addWaveform((waveform.sampleID, waveform.waveformData))
            self.dispatch?(action)
        }

        if response.hasUpload {
            let action = Action.uploadAck(response.upload.uploadID)
            self.dispatch?(action)
        }

        if response.hasProjectInfo {
            self.dispatch?(.setProjectInfo(response.projectInfo))
        }

        if response.hasUserStatus && response.userStatus.hasUser {
            self.dispatch?(.setUser(response.userStatus.user))
        }
        else if response.hasUserStatus {
            self.dispatch?(.clearUser)
        }

        if response.hasProjectSync {
            self.dispatch?(
                .setProjectSync(response.projectSync.projectID, response.projectSync.status)
            )
        }
    }
}

File: android/app/src/main/java/com/joenoel/bloop/state/ResponseMiddleware.kt
package com.joenoel.bloop.state

import bloop.Bloop

class ResponseMiddleware : AppMiddleware {
    override suspend fun execute(state: AppState, action: AppAction, dispatch: (AppAction) -> Unit) {
        if (action is AppAction.ReceivedResponse) {
            onResponse(action.response, dispatch)
        }
    }

    private fun onResponse(response: Bloop.Response, dispatch: (AppAction) -> Unit) {
        dispatch(AppAction.SetLastResponseText(response.toString()))

        if (response.hasProject()) {
            dispatch(AppAction.SetProject(response.project))
        }

        if (response.hasPlaybackState()) {
            dispatch(AppAction.SetPlaybackState(response.playbackState))
        }

        if (response.hasProgress()) {
            dispatch(AppAction.SetProgress(response.progress))
        }

        if (response.projectsList.isNotEmpty()) {
            dispatch(AppAction.SetProjects(response.projectsList))
        }

        if (response.cloudProjectsList.isNotEmpty()) {
            dispatch(AppAction.SetCloudProjects(response.cloudProjectsList))
        }

        if (response.error.isNotEmpty()) {
            dispatch(AppAction.AddError(response.error))
        }

        if (response.hasWaveform()) {
            val waveform = response.waveform
            dispatch(AppAction.AddWaveform(waveform.sampleId, waveform.waveformData))
        }

        if (response.hasUpload()) {
            dispatch(AppAction.UploadAck(response.upload.uploadId))
        }

        if (response.hasProjectInfo()) {
            dispatch(AppAction.SetProjectInfo(response.projectInfo))
        }

        if (response.hasUserStatus()) {
            if (response.userStatus.hasUser()) {
                dispatch(AppAction.SetUser(response.userStatus.user))
            } else {
                dispatch(AppAction.ClearUser)
            }
        }

        if (response.hasProjectSync()) {
            dispatch(AppAction.SetProjectSync(response.projectSync.projectId, response.projectSync.status))
        }

        if (response.hasPreferences()) {
            dispatch(AppAction.SetPreferences(response.preferences))
        }

        if (response.hasAudioDevices()) {
            dispatch(AppAction.SetAudioDevices(response.audioDevices))
        }

        if (response.hasAudioStatus()) {
            dispatch(AppAction.SetAudioStatus(response.audioStatus))
        }

        if (response.hasMidiDevices()) {
            dispatch(AppAction.SetMidiDevices(response.midiDevices))
        }
    }
}